### DISCUSSION TITLE
Compile exercises with and without answers on a website

### DISCUSSION THREAD

**Discussion by snhansen** [quarto-dev/quarto-cli#7830] [Q&A]
### Description

I'm looking for some advice on how to compile exercises with and without solutions for a website. Right now, I have the course material in the following structure

```
index.qmd
day1\day1.qmd [this creates listings of slides and exercises for day 1]
day2\day2.qmd [this creates listings of slides and exercises for day 2]
day1\slides\*.qmds
day1\exercises\*.qmds
day2\slides\*.qmds
day2\exercises\*.qmds
```

and I compile and include these via listings.  I'm aware that you can make the exercises documents so that, if you compile them with a certain parameter, e.g. `quarto render exercises.qmd -P par:TRUE`, then they're compiled with solutions (as code blocks). I wonder how I most easily can compile the exercises with solutions and include them somehow in the site. I can't seem to pass a parameter to a listing. Any suggestions?

**Comment by mcanouil** [comment]
- https://github.com/quarto-dev/quarto-cli/discussions/7127
- https://github.com/quarto-dev/quarto-cli/discussions/7314
- https://github.com/quarto-dev/quarto-cli/discussions/5823
- https://github.com/quarto-dev/quarto-cli/discussions/2422

Additionally, take a look at project profiles, see <https://quarto.org/docs/projects/profiles.html>.

**Reply by snhansen** [reply]
Thanks. Project profiles seems to be the way. But how would I go about rendering the whole project? Say I want to compile both exercises and exercises with solutions and include them on the website? Can you do profiles with listings?

**Reply by mcanouil** [reply]
Read the documentation about profiles, the answer is there.

**Reply by Marika-G** [reply]
Hello! I might have a solution to this problem, if I understood correctly what @snhansen wanted to do.

## Reformulation
From what I understand, the requirements are : 
1. Have only one .qmd file per exercise;
2. Being able to render it either as the assignment (w/o the answers) or with the answers;
3. Being able to have **both** versions in a single Quarto website.

Up to point 2, it can indeed easily be done with profiles and conditional content blocks. This works great if we want to have two distinct websites, one for assignments and another for answers. But if we want to create a website containing both versions, I have not found any way to do it with profiles.

## My suggestion

**However**, I found this could be done using [conditional content block](https://quarto.org/docs/authoring/conditional.html) (more precisely, with [metadata matching](https://quarto.org/docs/authoring/conditional.html)), [includes](https://quarto.org/docs/authoring/includes.html), and a  *little* compromise on file duplication.  This solution allows for : 
- a single file containing both the questions and the answers;
- having assignments AND answers pages in the same website;
- easy addition of pages to the website (e.g., as the semester goes, my personal use case for this).

I have not tested it in combination with listings or profiles, but it could probably be done to get even more flexibility.

>[!NOTE]
Be careful if using "freeze: auto" or "freeze: true", because changes made to the "main" code files will not affect the "answer files", therefore they will not be re-rendered.

### Working example

_quarto.yml
```yaml
project:
  type: website
    
execute:
  cache: true

website:
  title: "Very Interesting Class - Fall 2025"
  navbar:
    search: true
  sidebar:
    style: "docked"
    contents:
      - section: "Exercises"
        contents:
          - exercises_week1.qmd
          - exercises_week2.qmd

      - section: "Answers"
        contents:
          - answers_week1.qmd
```

index.qmd
```qmd
---
title: "Very Interesting Class - Fall 2025"
---

This is a course description.
```

exercises_week1.qmd
````qmd
---
title: "Week 1"
---

## First question 

Some clever problem for clever students.

:::{.content-hidden when-meta="answer"}
```{r}
# Here is some very useful computation for this problem,
# but useless in the answer
1 + 1
```
:::

:::{.content-visible when-meta="answer"}
## First answer

Some surprisingly wise answer.
:::
````

answers_week1.qmd
````qmd
---
answer: true
---

{{< include exercises_week1.qmd >}}
````

exercises_week2.qmd
```qmd
---
title: "Week 2"
---

## First question 

I'm running out of inspiration.

:::{.content-visible when-meta="answer"}
And I won't use LLMs for this.
:::
```

<img width="1098" height="442" alt="image" src="https://github.com/user-attachments/assets/13413854-6a45-471d-89ee-436b67ad5f26" />
<img width="1093" height="422" alt="image" src="https://github.com/user-attachments/assets/f0f4b17b-3304-4f45-8c0c-d8aa8d042ce3" />

**Comment by mcanouil** [comment]
FYI, there is a `live` Quarto extension for this: https://github.com/r-wasm/quarto-live
